### ISSUE TITLE
Can markers be calculated independently?

### ISSUE THREAD

**Issue by thkhavi** [DannyArends/CTLmapping#27]
Hi Dr. Arends,

Thank you for developing and making your CTLmapping code available. 

Based on the documentation in your dissertation and the JOSS manuscript, my (limited) understanding suggests that CTL for markers can be calculated independently. However, in trying to split up a matrix of markers to run CTL mapping in parallel across multiple nodes, I found that the results of CTLscan() for the whole set of markers were different from running CTLscan() on subsets of the markers.

Should the results for a single marker be independent of the other markers processed with it? I provided a test case below where using CTLscan() on two markers provides different results from running them individually.

Thanks for your help, and for making the implementation available!
Sandra Truong

```
library(qtl)
library(MASS)
library(ctl)

# Load Arabidopsis Thaliana data set
data(ath.metabolites)

# Define genotype and phenotype matrices
matrix_geno = ath.metab$genotypes
matrix_pheno = ath.metab$phenotypes

# Function mat_split from stackoverflow
# http://stackoverflow.com/questions/24299171/function-to-split-a-matrix-into-sub-matrices-in-r
mat_split <- function(M, r, c){
  nr <- ceiling(nrow(M)/r)
  nc <- ceiling(ncol(M)/c)
  newM <- matrix(NA, nr*r, nc*c)
  newM[1:nrow(M), 1:ncol(M)] <- M
  div_k <- kronecker(matrix(seq_len(nr*nc), nr, byrow = TRUE), matrix(1, r, c))
  matlist <- split(newM, div_k)
  N <- length(matlist)
  mats <- unlist(matlist)
  dim(mats)<-c(r, c, N)
  return(mats)
}

# Select the first marker and perform ctlscan
matrix_geno_one_marker_1 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 1)[,,1])
ctlscan_one_marker_1 = CTLscan(genotypes = matrix_geno_one_marker_1 , phenotypes = matrix_pheno)

# Select the second marker and perform ctlscan
matrix_geno_one_marker_2 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 1)[,,2])
ctlscan_one_marker_2 = CTLscan(genotypes = matrix_geno_one_marker_2 , phenotypes = matrix_pheno)

# Select the first and second markers and perform ctlscan
matrix_geno_two_markers_1_and_2 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 2)[,,1])
ctlscan_two_markers_1_and_2 = CTLscan(genotypes = matrix_geno_two_markers_1_and_2, phenotypes = matrix_pheno)

# Print phenotypes where the ctlscan results are different for the same marker but in different sets.
for(phenotype_i in 1:(ncol(matrix_pheno))){
  if (ctlscan_one_marker_1[[phenotype_i]]$ctl[1,] != ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[1,] && any(ctlscan_one_marker_1[[phenotype_i]]$ctl[1,] > 0.0)){
    print("ctlscan of marker 1 is not equivalent for scanning with one vs two markers for:")
    print(colnames(matrix_pheno)[phenotype_i])
    #print(ctlscan_one_marker_1[[phenotype_i]]$ctl[1,])
    #print(ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[1,])
  }
  if (ctlscan_one_marker_2[[phenotype_i]]$ctl[1,] != ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[2,] && any(ctlscan_one_marker_2[[phenotype_i]]$ctl[1,] > 0.0)){
    print("ctlscan of marker 2 is not equivalent for scanning with one vs two markers for:")
    print(colnames(matrix_pheno)[phenotype_i])
    #print(ctlscan_one_marker_2[[phenotype_i]]$ctl[1,])
    #print(ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[2,])
  }
}
```

**Comment by DannyArends** [comment]
I'll just add my reply email to the comment, (I was expecting Github to automatically add it)

Dear Sandra,

Yes you are correct ctl can be used on independent markers, however since multiple testing correction is done inside the algorithm results will be different if you run it on individual/single markers or on multiple markers, since the null hypothesis changes.

If you want I can send you a package which does not adjust p-values, this will yield similar (but inflated) p-values when scanning a single or multiple markers, and as such is very prone to over-fitting the model since a lot more tests are performed when scanning 1 compared to multiple markers in LD.

The best way of using ctl is to scan all markers and all phenotypes together in a single run,  this way multiple testing correction is then performed for the correct number of tests, and as such is the most safe way of performing ctl scans.

**Comment by DannyArends** [comment]
Another update,

I added a new parameter to the R function that disables adjustment for multiple testing, so that you're able to run it on a cluster (e.g. separated by chromosome).

You can install the newest version directly from Github using the devtools package, make sure you have the correct Rtools version installed (https://cran.r-project.org/bin/windows/Rtools/):

````R
    library(devtools)
    install_github("DannyArends/CTLmapping", subdir = "Rctl")
````

After that run your code, but in the CTLscan function set the parameter adjust to FALSE
this should produce identical p-values when running 1 marker or multiple markers. Please be aware that you do need to adjust p-values for multiple testing yourself (based on the number of markers and phenotypes tested)

Attached the updated code which after installing the package from Github doesn't give me any warnings.

````R
library(qtl)
library(MASS)
library(ctl)

# Load Arabidopsis Thaliana data set
data(ath.metabolites)

# Define genotype and phenotype matrices
matrix_geno = ath.metab$genotypes
matrix_pheno = ath.metab$phenotypes

# Function mat_split from stackoverflow
# http://stackoverflow.com/questions/24299171/function-to-split-a-matrix-into-sub-matrices-in-r
mat_split <- function(M, r, c){
  nr <- ceiling(nrow(M)/r)
  nc <- ceiling(ncol(M)/c)
  newM <- matrix(NA, nr*r, nc*c)
  newM[1:nrow(M), 1:ncol(M)] <- M
  div_k <- kronecker(matrix(seq_len(nr*nc), nr, byrow = TRUE), matrix(1, r, c))
  matlist <- split(newM, div_k)
  N <- length(matlist)
  mats <- unlist(matlist)
  dim(mats)<-c(r, c, N)
  return(mats)
}

# Select the first marker and perform ctlscan
matrix_geno_one_marker_1 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 1)[,,1])
ctlscan_one_marker_1 = CTLscan(genotypes = matrix_geno_one_marker_1 , phenotypes = matrix_pheno, adjust=FALSE)

# Select the second marker and perform ctlscan
matrix_geno_one_marker_2 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 1)[,,2])
ctlscan_one_marker_2 = CTLscan(genotypes = matrix_geno_one_marker_2 , phenotypes = matrix_pheno, adjust=FALSE)

# Select the first and second markers and perform ctlscan
matrix_geno_two_markers_1_and_2 = as.matrix(mat_split(matrix_geno, nrow(matrix_geno), 2)[,,1])
ctlscan_two_markers_1_and_2 = CTLscan(genotypes = matrix_geno_two_markers_1_and_2, phenotypes = matrix_pheno, adjust=FALSE)

# Print phenotypes where the ctlscan results are different for the same marker but in different sets.
for(phenotype_i in 1:(ncol(matrix_pheno))){
  if (ctlscan_one_marker_1[[phenotype_i]]$ctl[1,] != ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[1,] && any(ctlscan_one_marker_1[[phenotype_i]]$ctl[1,] > 0.0)){
    print("ctlscan of marker 1 is not equivalent for scanning with one vs two markers for:")
    print(colnames(matrix_pheno)[phenotype_i])
    #print(ctlscan_one_marker_1[[phenotype_i]]$ctl[1,])
    #print(ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[1,])
  }
  if (ctlscan_one_marker_2[[phenotype_i]]$ctl[1,] != ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[2,] && any(ctlscan_one_marker_2[[phenotype_i]]$ctl[1,] > 0.0)){
    print("ctlscan of marker 2 is not equivalent for scanning with one vs two markers for:")
    print(colnames(matrix_pheno)[phenotype_i])
    #print(ctlscan_one_marker_2[[phenotype_i]]$ctl[1,])
    #print(ctlscan_two_markers_1_and_2[[phenotype_i]]$ctl[2,])
  }
}
````

**Comment by thkhavi** [comment]
Hi Dr. Arends,

This is great, thank you! I really appreciate it.

I'll be sure to adjust the p-values for multiple testing in the post-CTL mapping steps. 

Sincerely,
Sandra Truong

**Comment by DannyArends** [comment]
Dear Sandra,

No problem, thanks for reporting it. I will update the version on CRAN as well, but that might take a little while. I think adding the extra parameter so that you can turn off the automatic p-value correction is a really good idea, this change will be indeed very useful for people trying to run it on multiple nodes / a cluster. I will close this issue, but if you have any questions in the future don't hesitate to open another issue.

Greetings and have a nice weekend,
Danny
